tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/join/ControlledWorkerBindJoinBase.java

```diff
@@ -0,0 +1,163 @@
+		this.scheduler = scheduler;
+	}
+
+	@Override
+	protected void handleBindings() throws Exception {
+		if (!(rightArg instanceof BoundJoinTupleExpr)) {
+			String msg = "Right argument is not an applicable expression for bind joins. Was: "
+					+ rightArg.getClass().getCanonicalName();
+			log.debug(msg);
+			throw new QueryEvaluationException(msg);
+		}
+
+		int nBindingsCfg = this.queryInfo.getFederationContext().getConfig().getBoundJoinBlockSize();
+		int totalBindings = 0; // the total number of bindings
+		TupleExpr expr = rightArg;
+
+		TaskCreator taskCreator = null;
+		Phaser currentPhaser = phaser;
+
+		int nBindings;
+		List<BindingSet> bindings;
+		while (!isClosed() && leftIter.hasNext()) {
+
+			// create a new phaser if there are more than 10000 parties
+			// note: a phaser supports only up to 65535 registered parties
+			if (currentPhaser.getRegisteredParties() >= 10000) {
+				currentPhaser = new Phaser(currentPhaser);
+			}
+
+			// determine the bind join block size
+			nBindings = getNextBindJoinSize(nBindingsCfg, totalBindings);
+
+			bindings = new ArrayList<>(nBindings);
+
+			int count = 0;
+			while (!isClosed() && count < nBindings && leftIter.hasNext()) {
+				var bs = leftIter.next();
+				if (taskCreator == null) {
+					taskCreator = determineTaskCreator(expr, bs);
+				}
+				bindings.add(bs);
+				count++;
+			}
+
+			totalBindings += count;
+
+			currentPhaser.register();
+			scheduler.schedule(taskCreator.getTask(new PhaserHandlingParallelExecutor(this, currentPhaser), bindings));
+		}
+
+		leftIter.close();
+
+		scheduler.informFinish(this);
+
+		if (log.isDebugEnabled()) {
+			log.debug("JoinStats: left iter of " + getDisplayId() + " had " + totalBindings + " results.");
+		}
+
+		phaser.awaitAdvanceInterruptibly(phaser.arrive(), queryInfo.getMaxRemainingTimeMS(), TimeUnit.MILLISECONDS);
+	}
+
+	@Override
+	public void handleClose() throws QueryEvaluationException {
+		try {
+			super.handleClose();
+		} finally {
+			// signal the phaser to close (if currently being blocked)
+			phaser.forceTermination();
+		}
+	}
+
+	/**
+	 * Return the {@link TaskCreator} for executing the bind join
+	 *
+	 * @param expr
+	 * @param bs
+	 * @return
+	 */
+	protected abstract TaskCreator determineTaskCreator(TupleExpr expr, BindingSet bs);
+
+	/**
+	 * Return the size of the next bind join block.
+	 *
+	 * @param configuredBindJoinSize the configured bind join size
+	 * @param totalBindings          the current process bindings from the intermediate result set
+	 * @return
+	 */
+	protected int getNextBindJoinSize(int configuredBindJoinSize, int totalBindings) {
+
+		/*
+		 * XXX idea:
+		 *
+		 * make nBindings dependent on the number of intermediate results of the left argument.
+		 *
+		 * If many intermediate results, increase the number of bindings. This will result in less remote SPARQL
+		 * requests.
+		 *
+		 */
+
+		return configuredBindJoinSize;
+	}
+
+	protected interface TaskCreator {
```

Comment:
hmm, moving this interface up to the base class seems to cause a class incompatbility:

```
Sep 06, 2024 12:24:10 PM japicmp.output.incompatible.IncompatibleErrorOutput warn
WARNING: Incompatibility detected: Requires semantic version level MAJOR: JApiMethod [oldMethod=org.eclipse.rdf4j.federated.evaluation.join.ControlledWorkerBoundJoin.determineTaskCreator(org.eclipse.rdf4j.query.algebra.TupleExpr,org.eclipse.rdf4j.query.BindingSet), newMethod=org.eclipse.rdf4j.federated.evaluation.join.ControlledWorkerBoundJoin.determineTaskCreator(org.eclipse.rdf4j.query.algebra.TupleExpr,org.eclipse.rdf4j.query.BindingSet), returnType=JApiReturnType [oldReturnTypeOptional=org.eclipse.rdf4j.federated.evaluation.join.ControlledWorkerBoundJoin$TaskCreator, newReturnTypeOptional=org.eclipse.rdf4j.federated.evaluation.join.ControlledWorkerBindJoinBase$TaskCreator, changeStatus=MODIFIED], getCompatibilityChanges()=[METHOD_RETURN_TYPE_CHANGED]]
Sep 06, 2024 12:24:10 PM japicmp.output.incompatible.IncompatibleErrorOutput warn
WARNING: Incompatibility detected: Requires semantic version level MAJOR: JApiMethod [oldMethod=org.eclipse.rdf4j.federated.evaluation.join.ControlledWorkerBoundJoin.setSubmitFirstResultImmediately(boolean), newMethod=n.a., returnType=JApiReturnType [oldReturnTypeOptional=void, newReturnTypeOptional=value absent, changeStatus=REMOVED], getCompatibilityChanges()=[METHOD_REMOVED]]
Sep 06, 2024 12:24:10 PM japicmp.output.incompatible.IncompatibleErrorOutput warn
WARNING: Incompatibility detected: Requires semantic version level MAJOR: JApiSuperclass [jApiClass=JApiClass [fullyQualifiedName=org.eclipse.rdf4j.federated.evaluation.join.ControlledWorkerBoundJoin, changeStatus=MODIFIED, compatibilityChanges=[METHOD_DEFAULT_ADDED_IN_IMPLEMENTED_INTERFACE]], oldSuperclass=org.eclipse.rdf4j.federated.evaluation.join.ControlledWorkerJoin, newSuperclass=org.eclipse.rdf4j.federated.evaluation.join.ControlledWorkerBindJoinBase, changeStatus=MODIFIED, compatibilityChanges=[SUPERCLASS_REMOVED]]
```

I will try to find a solution that remeains binary compatible

Reply:
you might create the interface of the same name in the parent class and extend this one from that? This should preserve compatibility.

Reply:
I have solved this now differently, by keeping the existing `ControlledWorkerBoundJoin` in place unmodifed, and marking it for removal. In the strategy the refactored `ControlledWorkerBindJoin` is now used.